internal/twirptest/service_test.go

```diff
@@ -1527,3 +1527,150 @@ func TestPanicFlushing(t *testing.T) {
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
+	// Associate the cancellable context we just created to the request
+	req = req.WithContext(ctx)
+	// cancel context
+	cancel()
+	w := httptest.NewRecorder()
+	server.ServeHTTP(w, req)
+
+	// validate error code
+	expectedErrCode := twirp.ServerHTTPStatusFromErrorCode(twirp.Canceled)
+	if w.Code != expectedErrCode {
+		t.Errorf("twirp ErrorCode expected to be %q, but found %q", expectedErrCode, w.Code)
+	}
+
+	// validate error message
+	respBytes, err := ioutil.ReadAll(w.Body)
+	if err != nil {
+		t.Fatalf("Could not even read bytes from response: %q", err.Error())
+	}
+
+	expectedErrMessage := "context canceled"
+	if !strings.Contains(string(respBytes), expectedErrMessage) {
+		t.Errorf("twirp client err has unexpected message %q, want %q", string(respBytes), expectedErrMessage)
+	}
+}
+
+func TestDeadlineExceededError(t *testing.T) {
+	// test case validate context deadline exceed error handling
+	var s Haberdasher
+	hooks := &twirp.ServerHooks{Error: func(ctx context.Context, err twirp.Error) context.Context {
+		if ctxErr := twirp.DeadlineExceeded; err.Code() != ctxErr {
+			t.Errorf("twirp ErrorCode expected to be %q, but found %q", ctxErr, err.Code())
+		}
+		if ctxErr := context.DeadlineExceeded.Error(); err.Msg() != ctxErr {
+			t.Errorf("twirp client err has unexpected message %q, want %q", err.Msg(), ctxErr)
+		}
+		return ctx
+	}}
+	server := NewHaberdasherServer(s, hooks)
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Millisecond)
+	defer cancel()
+	// Make a request, that will endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.Haberdasher/MakeHat", errReader{})
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
+	// Associate the cancellable context we just created to the request
+	req = req.WithContext(ctx)
+
+	// wait for deadline exceed timeout in context object
+	select {
+	case <-ctx.Done():
+		// ctx.Err() will have context.DeadlineExceeded
+	case <-time.After(1 * time.Second):
+		t.Fatalf("Waited too long for DeadlineExceeded ")
+	}
+	w := httptest.NewRecorder()
+
+	server.ServeHTTP(w, req)
+
+	// validate error code
+	expectedErrCode := twirp.ServerHTTPStatusFromErrorCode(twirp.DeadlineExceeded)
+	if w.Code != expectedErrCode {
+		t.Errorf("twirp ErrorCode expected to be %q, but found %q", expectedErrCode, w.Code)
+	}
+
+	// validate error message
+	respBytes, err := ioutil.ReadAll(w.Body)
+	if err != nil {
+		t.Fatalf("Could not even read bytes from response: %q", err.Error())
+	}
+
+	expectedErrMessage := "context deadline exceeded"
+	if !strings.Contains(string(respBytes), expectedErrMessage) {
+		t.Errorf("twirp client err has unexpected message %q, want %q", string(respBytes), expectedErrMessage)
+	}
+}
+
+func TestFailedToReadRequestBodyError(t *testing.T) {
```

Comment:
These three test cases have a lot of shared code. It's hard to see from a glance where they differ. Can you factor out the common behavior please?

It also gives a clear opportunity to add additional checks that should be true for all of these cases, including "because these are all client-initiated problems, the http status should be in the range [400,499]".

I find that a `testcase := func(differences) func(t *testing.T) {...}` at the start, followed by a bunch of calls to `t.Run("canceled-context", testcase(args))` is a helpful structure to share some behavior between tests.

Reply:
I merged three test cases into one. 

We dont have any client and not making any client call. Instead calling server method directly as unit test case. Still wondering how I can check [400, 499] error?

Instead [400, 499] we are checking specific error https://github.com/twitchtv/twirp/pull/294/files#diff-f9d0e6645e9e1b67ccaaceb3da4d91900b84e0cf1ed5d2b7a94f85a9f0423870R1609. Wondering if that is enough?